api/models/storagestats.go

```diff
@@ -0,0 +1,82 @@
+package models
+
+import (
+	"strconv"
+	"sync/atomic"
+
+	"github.com/grafana/metrictank/mdata/cache"
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+//go:generate msgp
+type StorageStats struct {
+	CacheMiss       uint32 `json:"executeplan.cache-miss.count"`
+	CacheHitPartial uint32 `json:"executeplan.cache-hit-partial.count"`
+	CacheHit        uint32 `json:"executeplan.cache-hit.count"`
+	ChunksFromTank  uint32 `json:"executeplan.chunks-from-tank.count"`
+	ChunksFromCache uint32 `json:"executeplan.chunks-from-cache.count"`
+	ChunksFromStore uint32 `json:"executeplan.chunks-from-store.count"`
+}
+
+func (ss *StorageStats) IncCacheResult(t cache.ResultType) {
+	switch t {
+	case cache.Hit:
+		atomic.AddUint32(&ss.CacheHit, 1)
+	case cache.HitPartial:
+		atomic.AddUint32(&ss.CacheHitPartial, 1)
+	case cache.Miss:
+		atomic.AddUint32(&ss.CacheMiss, 1)
+	}
+}
+
+func (ss *StorageStats) IncChunksFromTank(n uint32) {
+	atomic.AddUint32(&ss.ChunksFromTank, n)
+}
+func (ss *StorageStats) IncChunksFromCache(n uint32) {
+	atomic.AddUint32(&ss.ChunksFromCache, n)
+}
+func (ss *StorageStats) IncChunksFromStore(n uint32) {
+	atomic.AddUint32(&ss.ChunksFromStore, n)
+}
+
+// Add adds a to ss. Note that a is presumed to be "done" (read unsafely)
```

Comment:
While it would only make sense to try and add `a` once it is 'done', it would be unwise/unsafe to presume that is always true. 
We should just use atomic operations to read `a` to avoid the possibility of a data race.  eg
```
atomic.AddUint32(&ss.CacheHit, atomic.LoadUint32(&a.CacheHit)))
```

The call to `Add()` itself wont be atomic (i.e. individual counters might increase between when the first counter from `a` is read and when the last one is), but that wont have any impact on our use case.

Reply:
i have pondered the same thing.
I considered it the programmer's responsibility to make sure `a` is "done". i agree using atomics for `a` is safer, but at the same time, it would mask issues should they appear: it doesn't make sense to compute the aggregate stats if we're not done yet generating the stats.

I think this is in line with best practices everywhere in go code: whenever calling a function it's the programmers responsibility to make sure there is no unsafe data access. this is true for stdlib and many 3rd party libraries.

Reply:
> it would mask issues should they appear

The main issue we need to avoid is panics.  Panics are bad and we should avoid them were ever possible.  Wrapping the the reads in atomic.LoadUint32 avoids panics.  It doesn't avoid other issues that may result from users adding stats that are being updated and that is fine.  The comment about `a` being done should still stay.

> I think this is in line with best practices everywhere in go code:

As a general statement that is true.  But a more specific "best practice" is "When using atomic, all reads and writes need to use atomic".  We have run into lots of race issues because this simple rule hasn't been followed.   So, we either need to always use atomic operations or not use them at all.

Reply:
makes sense. actually the way the memory model works is you shouldn't mix atomic access with non-atomic ones, even when you know you're not accessing concurrently. Go's memory model provides no guarantees if you don't use atomics and the compiler/runtime actually take advantage of this to do memory optimizations behind the scenes.
it's so easy to forget this. funny, because i recently pointed it out somewhere else as well.

i'll fix